server/src/main/java/com/objectcomputing/checkins/services/checkins/CheckInServicesImpl.java

```diff
@@ -1,86 +1,167 @@
+        Boolean isAdmin = securityService!=null ? securityService.hasRole(RoleType.Constants.ADMIN_ROLE) : false;
+
+        if(checkIn.getId()!=null) {
             throw new CheckInBadArgException(String.format("Found unexpected id for checkin  %s", checkIn.getId()));
-        } else if(memberId.equals(pdlId)){
+        } else if(memberId.equals(pdlId)) {
             throw new CheckInBadArgException(String.format("Team member id %s can't be same as PDL id", checkIn.getTeamMemberId()));
-        } else if(!memberRepo.findById(memberId).isPresent()){
-            throw new CheckInBadArgException(String.format("Member %s doesn't exists", memberId));
-        } else if(chkInDate.isBefore(LocalDate.EPOCH) || chkInDate.isAfter(LocalDate.MAX)) {
+        } else if(memberRepo.findById(memberId).isEmpty()) {
+            throw new CheckInBadArgException(String.format("Member %s doesn't exist", memberId));
+        } else if(!pdlId.equals(memberRepo.findById(memberId).get().getPdlId())) {
+            throw new CheckInBadArgException(String.format("PDL %s is not associated with member %s", pdlId, memberId));
+        } else if(chkInDate.isBefore(Util.MIN) || chkInDate.isAfter(Util.MAX)) {
             throw new CheckInBadArgException(String.format("Invalid date for checkin %s",memberId));
+        } else if(!isAdmin) {
+            if(!currentUser.getId().equals(checkIn.getTeamMemberId()) && !currentUser.getId().equals(checkIn.getPdlId())) {
+                // Limit create to subject of check-in, PDL of subject and Admin
+                throw new CheckInBadArgException(String.format("Member %s is unauthorized to do this operation", currentUser.getId()));
+            }
         }
-        checkInRet = checkinRepo.save(checkIn);
-        }
-            return checkInRet ;
-    }
 
+        return checkinRepo.save(checkIn);
+    }
 
     @Override
     public CheckIn read(@NotNull UUID id) {
 
-        return checkinRepo.findById(id).orElse(null);
+        String workEmail = securityService!=null ? securityService.getAuthentication().get().getAttributes().get("email").toString() : null;
+        MemberProfile currentUser = workEmail!=null? currentUserServices.findOrSaveUser(null, workEmail) : null;
+        Boolean isAdmin = securityService!=null ? securityService.hasRole(RoleType.Constants.ADMIN_ROLE) : false;
+
+        CheckIn result = checkinRepo.findById(id).orElse(null);
+
+        if (result == null) {
+            throw new CheckInBadArgException(String.format("Invalid checkin id %s", id));
+        } else if(!isAdmin) {
+            if(!currentUser.getId().equals(result.getTeamMemberId()) && !currentUser.getId().equals(result.getPdlId())) {
+                // Limit read to Subject of check-in, PDL of subject and Admin
+                throw new CheckInBadArgException(String.format("Member %s is unauthorized to do this operation", currentUser.getId()));
```

Comment:
The ID isn't very meaningful to a user. I would nix the id here and just say something like "You are not authorized to perform this operation". It probably wouldn't hurt for us to have some AOP based logging that logs any thrown exceptions and records things like who the user was the method name and what the inputs to the method were.

Reply:
To be clear, that wasn't intended to indicate you should implement AOP based logging in this PR.  Just a thought about something we might want to consider adding.

Reply:
Sounds good, I've created a new story to implement AOP based logging. For the current PR, I've updated error messages.

New Issue - https://github.com/oci-labs/check-ins/issues/309